### PR TITLE
Padding for boolean formatting

### PR DESCRIPTION
**Pull request by xclerc**
I am not sure this is tremdendously useful, but I just
found mildly annoying to be unable to use padding when
printing boolean values through the `%B` specifier.

### CODE REVIEW

**Comment by damiendoligez** [comment]
Please don't include boostrap binaries in your PR. We have several good reasons to cut down the number of bootstrap binaries present in the git history.

**Comment by xclerc** [comment]
Sorry for the mistake; I have seen bootstrap binaries in other PR and thought it was OK.

I have a question, though: is the CI doing a bootstrap cycle or is there a way to tell it to
do so? (note: when working on formats used in the compiler, the bootstrap is tedious)

**Comment by shindere** [comment]
No, currently CI does not bootstrap and there is no way to request it to
do so.

**Comment by xclerc** [comment]
After a live discussion with @damiendoligez and @shindere, it seems
that the way forward would be to add a GitHub label (e.g. `needs-bootstrap`),
and to have the CI look for this flag to trigger a bootstrap where necessary.
We could then get rid of bootstrap binaries in pull requests.

Pull requests that entail a bootstrap should also be split (parts after/before),
so that the bootstrap can actually happen. It seems fragile (at best) to rely
on two successive commits (as commits are usually added to a PR during
the review process). What kind of meta data (that can be read by the CI)
can we attach to a pull request? The sha1 of a commit could be enough.

**Comment by shindere** [comment]
Xavier Clerc (2017/09/12 02:22 -0700):
> After a live discussion with @damiendoligez and @shindere, it seems
> that the way forward would be to add a GitHub label (e.g. `needs-bootstrap`),
> and to have the CI look for this flag to trigger a bootstrap where necessary.
> We could then get rid of bootstrap binaries in pull requests.
> 
> Pull requests that entail a bootstrap should also be split (parts after/before),
> so that the bootstrap can actually happen. It seems fragile (at best) to rely
> on two successive commits (as commits are usually added to a PR during
> the review process). What kind of meta data (that can be read by the CI)
> can we attach to a pull request? The sha1 of a commit could be enough.

About this very last point: if things are SHA1-based, wouldn't that be a
problem when PRs get rebased and the SHA1 change?

**Comment by xclerc** [comment]
It could indeed be a problem, and actually makes be realize
that the issue may be more complex that I originally thought:
what I described may be approximately reasonable for a mere
PR, but not for trunk. Unless the merge of a PR involving a
bootstrap actually updates the bootstrap binaries. It is possible
to have some kind of hook to do so?

(Also, this discussion may be moved to another place, as it is
clearly not specific to this very PR.)

**Comment by xclerc** [comment]
@trefis suggested offline to have an empty commit with a *magic* name
to mark where/when the bootstrap should occur. It is clearly nicer, as it
is more independent from GitHub, rebase-proof, and paves the way to
more complex schemes. The empty commit could also probably be
squashed when merging the PR into trunk.

**Comment by alainfrisch** [comment]
> to have the CI look for this flag to trigger a bootstrap where necessary.
We could then get rid of bootstrap binaries in pull requests.

I'm not sure to understand how this would work.  Certainly, some PR might require bootstrap to happen in the middle of commits.

Also, I'm not sure that it is good (or always possible) that CI pushes to the PR branch itself.

What about simply requiring CI to double-check bootstrap commits, i.e. those touching boot/ocamlc or boot/ocamllex?  For such commits, it would pick the content of boot/ from the parent commit, run the boostrap, and check that the result is identifical to the one in the current commit.

For this to work, one needs the bootstrap process to produce fully reproducible builds (across machines, and even preferably across OS).  I'm not sure exactly how far we are for this goal, but there are at least:

 - https://caml.inria.fr/mantis/view.php?id=6200
 - https://caml.inria.fr/mantis/view.php?id=6199

**Comment by xclerc** [comment]
I don't think we need the CI to push to the PR branch. I expect the CI to
be informed that a bootstrap should occur at a given point, and follow its
current process, only augmented by the bootstrap steps. The CI would
then return the usual OK/KO status as of today -- nothing being pushed
back to the PR.

Then, my understanding is that bootstrap binaries would be updated in the
repository when merging the PR. Probably through a git hook, although I
am not entirely sure it is possible.

That being said, your double-check workflow seems doable, even if I see
at least two downsides:
  - bootstrap binaries are still pushed to the repository;
  - achieving fully reproducible builds may require more work (*).

(*) a temporary measure may be to require people pushing PRs needing
a bootstrap to use a fixed set of configure switches (including, of course,
prefix) -- in these days of virtualization it seems quite possible, and such
PRs are in limited number anyway.

**Comment by xclerc** [comment]
(We may want to address [MPR #7600](https://caml.inria.fr/mantis/view.php?id=7600) along the way.)

**Comment by xclerc** [comment]
My understanding is that we should continue discussing the way
to handle bootstrap in PR/CI independently of the status of this
particular PR.

Is someone available to review the changes? @gasche maybe?

File: stdlib/camlinternalFormat.ml [hunk 1]
```diff
@@ -1500,8 +1505,8 @@ fun k o acc fmt -> match fmt with
     make_int_padding_precision k o acc rest pad prec convert_int64 iconv
   | Float (fconv, pad, prec, rest) ->
     make_float_padding_precision k o acc rest pad prec fconv
-  | Bool rest ->
-    fun b -> make_printf k o (Acc_data_string (acc, string_of_bool b)) rest
+  | Bool (pad, rest) ->
+    make_string_padding k o acc rest pad string_of_bool
```
Comment (gasche) [inline stdlib/camlinternalFormat.ml]: I would be a bit nicer to rename this `make_string_padding` function into `make_padding`, given that it is not string-specific anymore. But the current version is correct (it so happens that the string function had just the right amount of genericity to support booleans as well).

**Review by gasche — APPROVED** [review]
The code looks fine to me. I made a very minor nitpick in one comment, and I have another potential nitpick to make: given that the Bool case is now very similar to the String (or Caml_string) case for many functions, as they both only take a padding modifier, it would have made sense to move the Bool case directly after the last String case to keep looking-alike code portion close together. But the current way is also ok.

Some of the parts that handle padding/precision may benefit from a refactoring to avoid duplication of logic, but this is independent from this PR.

File: stdlib/camlinternalFormat.ml (hunk 1 above)
Comment (xclerc) [reply stdlib/camlinternalFormat.ml]: Renamed.

**Comment by xclerc** [comment]
Two more commits (`b492e91` and `f97d901`) in order to
get rid of (deprecated) `%b` specifiers.

**Comment by damiendoligez** [comment]
> What kind of meta data (that can be read by the CI) can we attach to a pull request?

I'm not a CI expert but couldn't we just put some kind of annotation in the message of the commit that comes right before the bootstrap?

>Unless the merge of a PR involving a bootstrap actually updates the bootstrap binaries.

We could do this by hand if there aren't too many PRs that need bootstrapping.

**Comment by dra27** [comment]
We had discussed the bootstrap issue for PRs in February (I made the mistake of saying I'd look into it further, but I haven't had a chance)... my memory was that the conclusion was that we would at some version move towards requiring that the CI can verify that a bootstrap is both needed and works (which would mean that the kind of commits Alain refers to would have to be split into two parts and should not be squashed when the PR itself is merged). This would also move towards a slightly more secure ideal that the repeatability of bootstrapping can be verified (i.e. that the bootstrap bytecode image really corresponds to the code).

I expect that trying to do this for 4.06 will be tricky :wink:

**Comment by alainfrisch** [comment]
> which would mean that the kind of commits Alain refers to would have to be split into two parts

Actualy, I'm no longer sure we would need to enforce such splitting.  Any commit that touches the bootstrap executables would just need to be verified (starting from the previous version of the bootstrap executables and doing a bootstrap cycle).

**Comment by dra27** [comment]
I was under the impression that if you:
1. Add a new primitive
2. Bootstrap
3. Use the new primitive in the bootstrap images
4. Bootstrap
5. Commit all changes

That you have an unrepeatable bootstrap - or is that not correct?

**Comment by alainfrisch** [comment]
> That you have an unrepeatable bootstrap - or is that not correct?

I don' know what you mean by unrepeatable.  I think that indeed, you cannot squash all commits together.  But I don't see anything wrong with squashing the bootstrap commits with the one just above them.  You have the same information available.

**Comment by gasche** [comment]
I approve the change, but I don't know the bootstrap machinery enough to merge this, and don't have the time to handle it this week. If someone else is interested in merging, please feel free.

**Comment by xclerc** [comment]
I think the CI issue should be handled in a dedicated PR, and should not
block the merge.

**Comment by lpw25** [comment]
This looks correct to me, and there's nothing particular about its bootstrap commits compared to other patches in the past, so I'm merging it.

**Pull request merged**